# BSA_crypto_analysis
A repository of tools used for Bulk segregant analysis in Cryptococcus neoformans

A R package for BSA analysis is available [here](https://github.com/cmatKhan/BSA). To cite the package or this analysis, see the end of this document.

## Citation
Daniel Paiva Agustinho, Holly Leanne Brown, Guohua Chen, Elizabeth Anne Gaylord, Jennifer Geddes-McAlister, Michael Richard Brent, Tamara Lea Doering,
Unbiased discovery of natural sequence variants that influence fungal virulence,
Cell Host & Microbe,
2023,
,
ISSN 1931-3128,
https://doi.org/10.1016/j.chom.2023.10.002.
(https://www.sciencedirect.com/science/article/pii/S1931312823004006)
Abstract: Summary
Isolates of Cryptococcus neoformans, a fungal pathogen that kills over 112,000 people each year, differ from a 19-Mb reference genome at a few thousand up to almost a million DNA sequence positions. We used bulked segregant analysis and association analysis, genetic methods that require no prior knowledge of sequence function, to address the key question of which naturally occurring sequence variants influence fungal virulence. We identified a region containing such variants, prioritized them, and engineered strains to test our findings in a mouse model of infection. At one locus, we identified a 4-nt variant in the PDE2 gene that occurs in common laboratory strains and severely truncates the encoded phosphodiesterase. The resulting loss of phosphodiesterase activity significantly impacts virulence. Our studies demonstrate a powerful and unbiased strategy for identifying key genomic regions in the absence of prior information and provide significant sequence and strain resources to the community.
Keywords: Cryptococcus neoformans; bulked segregant analysis; natural sequence variants; Pde2


